File: julia/dnd-character/dnd-character.jl
function modifier(ability)
    # Int(floor((ability - 10) / 2))
    fld(ability - 10, 2)
end

function ability()
    rolls = rand(1:6, 4)
    sum(rolls) - minimum(rolls)
end

mutable struct DNDCharacter
    strength::Int
    dexterity::Int
    constitution::Int
    intelligence::Int
    wisdom::Int
    charisma::Int
    hitpoints::Int

    function DNDCharacter()
        abilities = [ability() for i in 1:6]
        push!(abilities, 10 + modifier(abilities[3]))
        new(abilities...)
    end
end

File: ruby/dnd-character/dnd_character.rb
class DndCharacter
  def self.modifier(points)
    # silly OO arithmetic
    points.-(10).div(2)
  end

  attr_reader :strength, :dexterity, :constitution, :intelligence, :wisdom, :charisma, :hitpoints

  def initialize
    rand = Random.new

    ability = Proc.new do
      dice = 4.times.collect { rand.rand(1..6) }
      dice.sum - dice.min
    end

    @strength = ability.call
    @dexterity = ability.call
    @constitution = ability.call
    @intelligence = ability.call
    @wisdom = ability.call
    @charisma = ability.call

    @hitpoints = self.class.modifier(@constitution) + 10
  end
end
